### PR TITLE
fix(identity): correctly follow extract-expand for HKDF

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

As @mxinden pointed out in https://github.com/libp2p/rust-libp2p/pull/4554#discussion_r1344747938, we were not correctly following the HKDF steps of extract and expand.

## Notes & open questions

I intend to directly release this once it is merged.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: identity/src/keypair.rs [hunk 1]
```diff
@@ -360,19 +360,21 @@ impl Keypair {
...
-        return Some(
-            hkdf::Hkdf::<sha2::Sha256>::extract(None, &[domain, &self.secret()?].concat())
-                .0
-                .into(),
-        );
+        )))]
+        return None;
+
+        let hkdf =
+            hkdf::Hkdf::<sha2::Sha256>::from_prk(&self.secret()?).expect("secret to have 32 bytes");
```
Comment (mxinden) [inline identity/src/keypair.rs]: Why `from_prk` and not the default `new`?

From the documentation:

> If you already have a strong key to work from (uniformly-distributed and long enough), you can save a tiny amount of time by skipping the extract step. In this case, you pass a Pseudo-Random Key (PRK) into the [Hkdf::from_prk](https://docs.rs/hkdf/latest/hkdf/struct.Hkdf.html#method.from_prk) constructor, then use the resulting [Hkdf](https://docs.rs/hkdf/latest/hkdf/struct.Hkdf.html) object as usual.

Performance is irrelevant here. Why not follow the default way of doing things?

Again, coming from a novice.

Comment (thomaseizinger) [reply identity/src/keypair.rs]: > > If you already have a strong key to work from (uniformly-distributed and long enough), you can save a tiny amount of time by skipping the extract step. In this case, you pass a Pseudo-Random Key (PRK) into the [Hkdf::from_prk](https://docs.rs/hkdf/latest/hkdf/struct.Hkdf.html#method.from_prk) constructor, then use the resulting [Hkdf](https://docs.rs/hkdf/latest/hkdf/struct.Hkdf.html) object as usual.
> 
> Performance is irrelevant here. Why not follow the default way of doing things?

I didn't do it for performance reasons but I read that as the recommended way. If you have a strong key, we can use that. I think not doing that would be the odd thing, no? If I'd come across this code and see that `new` is used despite a strong key, I'd wonder why the author would think that the key wasn't strong enough despite being the node's private key.

Comment (mxinden) [reply identity/src/keypair.rs]: Do the relatively short ed25519 keys count as strong in this case? What is "long enough" here?

Comment (thomaseizinger) [reply identity/src/keypair.rs]: It must be at least as long as the output which is 32b for sha256. That is what it is in the source code at least.

Happy to use `new` as well. I didn't anticipate this to steer up this much discussion :)

**Review by mxinden — APPROVED** [review]
Fine with either `from_prk` and `new`. Thanks for explaining.

**Comment by mxinden** [comment]
For the record, published as `libp2p-identity` `v0.2.5`: https://crates.io/crates/libp2p-identity/0.2.5

Thank you @thomaseizinger.

**Pull request merged**